### PR TITLE
CNF-16104: Validate resources from hwmgr against nodepool request

### PR DESCRIPTION
**Pull request by abraham2512**
The goal of this PR is to validate the created Resource Group from the hardware manager against the nodepool request and to ensure all the nodes from the nodepool request are correctly satisified.

### CODE REVIEW

File: internal/controller/utils/nodepool_utils.go [hunk 1]
```diff
@@ -195,3 +197,49 @@ func NodepoolRemoveFinalizer(
...
+func ValidateNodepoolWithResourceSelector(
+	ctx context.Context,
+	nodepool *hwmgmtv1alpha1.NodePool,
+	resourceSelector map[string]api.RhprotoResourceSelectorGetResponse,
+) error {
+
+	for _, node := range nodepool.Spec.NodeGroup {
+		nodeName := node.NodePoolData.Name
+
+		if resource, exists := resourceSelector[nodeName]; exists {
```
Comment (rauhersu) [inline internal/controller/utils/nodepool_utils.go:210]: As a general comment, if we can't ensure all the pointers are not 'nil', maybe we might add a check previous to get the value to make this code more robust on de-referencing (*resource.X). Cc: @donpenney

File: internal/controller/utils/nodepool_utils.go [hunk 2]
```diff
@@ -20,6 +20,8 @@ import (
 	"context"
 	"fmt"
 
+	// is this import allowed here?
+	api "github.com/openshift-kni/oran-hwmgr-plugin/adaptors/dell-hwmgr/generated"
```
Comment (donpenney) [inline internal/controller/utils/nodepool_utils.go:24]: No, this code doesn't belong in the common utils. It should only be under adaptors/dell-hwmgr

Comment (abraham2512) [reply internal/controller/utils/nodepool_utils.go]: moved under adaptors/dell-hwmgr

File: internal/controller/utils/nodepool_utils.go [hunk 3]
```diff
@@ -195,3 +197,49 @@ func NodepoolRemoveFinalizer(
...
+	ctx context.Context,
+	nodepool *hwmgmtv1alpha1.NodePool,
+	resourceSelector map[string]api.RhprotoResourceSelectorGetResponse,
+) error {
+
+	for _, node := range nodepool.Spec.NodeGroup {
+		nodeName := node.NodePoolData.Name
+
+		if resource, exists := resourceSelector[nodeName]; exists {
+			if node.NodePoolData.HwProfile != *resource.ResourceProfileId {
```
[the remark this answers is not in the document]
Comment (abraham2512) [reply internal/controller/utils/nodepool_utils.go]: Removed this check.

File: internal/controller/utils/nodepool_utils.go [hunk 4]
```diff
@@ -195,3 +197,49 @@ func NodepoolRemoveFinalizer(
...
+					nodeName, float32(node.Size), *resource.NumResources)
+			}
+			if node.NodePoolData.ResourcePoolId != *resource.RpId {
+				return fmt.Errorf("invalid resource pool id for node %s\n expected: %s found: %s",
+					nodeName, node.NodePoolData.ResourcePoolId, *resource.RpId)
+			}
+			// Find value for key role in included labels
+			role := ""
+			if resource.Filters != nil && resource.Filters.Include != nil {
+				for _, label := range *resource.Filters.Include.Labels {
```
[the remark this answers is not in the document]
Comment (abraham2512) [reply internal/controller/utils/nodepool_utils.go]: Removed this check.

File: adaptors/dell-hwmgr/nodepool.go [hunk 5]
```diff
@@ -193,6 +193,16 @@ func (a *Adaptor) HandleNodePoolProcessing(
 
 	// TODO: Need to add validation to ensure the rg satisfies the nodepool
 
+	// WIP: Perform Validation
+	a.Logger.InfoContext(ctx, "Validating created resource group with nodepool")
+
+	if err := utils.ValidateNodepoolWithResourceSelector(ctx, nodepool, *rg.ResourceSelectors); err != nil {
```
[the remark this answers is not in the document]
Comment (abraham2512) [reply adaptors/dell-hwmgr/nodepool.go]: Update NodePool Status as Failed on err

File: internal/controller/utils/nodepool_utils.go [hunk 6]
```diff
@@ -195,3 +197,49 @@ func NodepoolRemoveFinalizer(
...
+					if label.Key != nil && *label.Key == "role" {
+						role = *label.Value
+						break
+					}
+				}
+			}
+			// Comparing role to node name
+			if node.NodePoolData.Name != role {
```
[the remark this answers is not in the document]
Comment (abraham2512) [reply internal/controller/utils/nodepool_utils.go]: Excluded this check

**Comment by donpenney** [comment]
FYI: For a "work-in-progress" PR, you can prefix the PR title with "WIP: <title>", and prow will automatically add a /hold. Similarly, you can create the PR as "draft" initially. You can also manually add a /hold while it's in-progress, removing the hold once ready

File: adaptors/dell-hwmgr/hwmgrclient/client.go [hunk 7]
```diff
@@ -336,3 +336,28 @@ func (c *HardwareManagerClient) GetSecret(ctx context.Context, secretKey string)
 
 	return response.JSON200, nil
 }
+
+// ValidateNodepoolwithResourceSelector validates the hardware manager resource group data with nodepool
+func (c *HardwareManagerClient) ValidateNodepoolWithResourceSelector(
```
Comment (donpenney) [inline adaptors/dell-hwmgr/hwmgrclient/client.go:341]: I would suggest renaming this to `ValidateResourceGroup`, aligning with the `GetResourceGroup` function, and checking passing the full `rg *hwmgrapi.RhprotoResourceGroupObjectGetResponseBody` as parameter

File: internal/controller/utils/nodepool_utils.go (hunk 1 above)
Comment (abraham2512) [reply internal/controller/utils/nodepool_utils.go]: Added safety checks to ensure de-referencing is safe. Returns error with `missing resource..`  for de-reference fail.

File: adaptors/dell-hwmgr/hwmgrclient/client.go (hunk 7 above)
Comment (abraham2512) [reply adaptors/dell-hwmgr/hwmgrclient/client.go]: Done, passing the full response body instead and added de-reference error check for *resourceGroup.ResourceSelectors

**Comment by abraham2512** [comment]
Node provisioned and installing after updates

File: adaptors/dell-hwmgr/hwmgrclient/client.go [hunk 8]
```diff
@@ -336,3 +336,44 @@ func (c *HardwareManagerClient) GetSecret(ctx context.Context, secretKey string)
...
+
+// ValidateResourceGroup validates the hardware manager resource group data with nodepool
+func (c *HardwareManagerClient) ValidateResourceGroup(
+	ctx context.Context,
+	nodepool *hwmgmtv1alpha1.NodePool,
+	resourceGroup hwmgrapi.RhprotoResourceGroupObjectGetResponseBody,
+) error {
+	if *resourceGroup.ResourceSelectors != nil {
+		resourceSelector := *resourceGroup.ResourceSelectors
+		for _, node := range nodepool.Spec.NodeGroup {
```
Comment (donpenney) [inline adaptors/dell-hwmgr/hwmgrclient/client.go:348]: The iterator here isn't a node, it's a nodegroup.
```suggestion
		for _, nodegroup := range nodepool.Spec.NodeGroup {
```

File: adaptors/dell-hwmgr/hwmgrclient/client.go [hunk 9]
```diff
@@ -336,3 +336,44 @@ func (c *HardwareManagerClient) GetSecret(ctx context.Context, secretKey string)
...
+			nodeName := node.NodePoolData.Name
```
Comment (donpenney) [inline adaptors/dell-hwmgr/hwmgrclient/client.go:349]: This isn't the node name. This is the nodegroup name
```suggestion
			nodegroupName := node.NodePoolData.Name
```

File: adaptors/dell-hwmgr/hwmgrclient/client.go [hunk 10]
```diff
@@ -336,3 +336,44 @@ func (c *HardwareManagerClient) GetSecret(ctx context.Context, secretKey string)
...
 	return response.JSON200, nil
 }
+
+// ValidateResourceGroup validates the hardware manager resource group data with nodepool
+func (c *HardwareManagerClient) ValidateResourceGroup(
+	ctx context.Context,
+	nodepool *hwmgmtv1alpha1.NodePool,
+	resourceGroup hwmgrapi.RhprotoResourceGroupObjectGetResponseBody,
+) error {
+	if *resourceGroup.ResourceSelectors != nil {
```
Comment (donpenney) [inline adaptors/dell-hwmgr/hwmgrclient/client.go:346]: You need to first check whether `resourceGroup.ResourceSelectors` is nil before dereferencing it.
```suggestion
	if resourceGroup.ResourceSelectors != nil && *resourceGroup.ResourceSelectors != nil {
```

**Comment by abraham2512** [comment]
Cluster validated and installed with latest changes

**Pull request merged**